### ISSUE TITLE
Support of asynchronous bidirectional streaming

### ISSUE THREAD

**Issue by mr2dark** [vmagamedov/grpclib#56]
I would call the current implementation (`grpclib==0.2.2rc2`) of `grpclib.client.StreamStreamMethod.__call__` rather synchronous.
The input messages are considered as of known size and the output items are gathered in a list before return.
For purposes like realtime speech recognition both the mentioned features are not applicable.
I actually managed to workaround this working directly with the stream being returned by `grpclib.client.StreamStreamMethod.open()` method.
Please see the illustrative code snippet below.
Currently the code snippet uses the separate task for sending messages while processing received messages in the main body but this can be done vice versa as well. 

And this code snippet fails with `RuntimeError('Concurrent call detected')` because `Wrapper` supports only one task to be cancelled. Actually `Wrapper` doesn't cancel anything but `DeadlineWrapper` does.
For testing purposes I removed the `Wrapper`'s check for concurrent calls and everything works fine. 

So in order for gprclib to properly support asynchronous bidirectional streaming I propose to implement the following:
1. Remove the concurrent call check for `Wrapper` as it's not used at this level anyway
1. Add support for several tasks to be cancelled via `DeadlineWrapper`
1. Or add support for several tasks to be taken into account in `Wrapper`

Also you can consider `grpclib.client.StreamStreamMethod.__call__` to be implemented in the way closer to logic presented in the snippet below.
1. Make `grpclib.client.StreamStreamMethod.__call__` accepting an asyncronous iterator for messages to be sent and converting it to a task
1. Make it returning the stream (not the list) and the message sending task so it can be managed later (cancelled, awaited etc.)

I do understand that this can result in rather overcomplicated API so I don't insist on this.
 
```python
loop = asyncio.get_event_loop()
channel = grpclib.client.Channel(server, port, loop=loop, ssl=True)
stub = ServiceStub(channel)

async def stream_sender(stream, ...):
    data = await <get more data>
    while data is not None:
        await stream.send_message(Request(content=data))
        data = await <get more data>
    await stream.end()

async with stub.Recognize.open() as stream:
    send_task = asyncio.ensure_future(stream_sender(stream, ...))
    async for response in stream:
        <process the response>
    await send_task
```

**Comment by vmagamedov** [comment]
> Actually `Wrapper` doesn't cancel anything but `DeadlineWrapper` does.

`Wrapper` cancels current task when the other side cancels current RPC call, `DeadlineWrapper` just adds timeouts/deadlines support.

In your example above `send_task` is not supervised properly. If your code gets cancelled, `send_task` may continue to work indefinitely. Currently in `asyncio` we don't have primitives to manage child tasks properly, like in `trio`: https://trio.readthedocs.io/en/latest/reference-core.html#nurseries-and-spawning, so this requires extra knowledge and extra work to write correct code.

It is possible to write stream-stream method handlers without creating child tasks in most cases, and code will be simpler. But I agree that in some cases we will want to send and receive messages truly independently in different tasks. So I think that `Wrapper` class should be fixed to support such cases.

Other intentions:
- `Method.__call__()` methods are made to be simple and safe
- `async with Method.open()` context managers are made to support more complex cases and also to be safe

Safety for me is when resources are cleaned correctly, and errors are not hidden.

That's why `Method.__call__()` doesn't return async iterators, because it is not possible to clean them properly, when your code gets cancelled somewhere in the middle of iteration. The only way is to use context managers. And accepting async iterators as an argument requires to work with child tasks inside `grpclib` and this also leads to complexity.

Maybe new example should be added to show stream-steam method type usage without using child tasks.

**Comment by vmagamedov** [comment]
Released `grpclib==0.2.2rc3` with a fix for the `Wrapper` class, can you confirm that this fix solves your issue?

**Comment by vmagamedov** [comment]
no response, please reopen if I missed something important
